leaderelection/leaderelection.go

```diff
@@ -0,0 +1,147 @@
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package leaderelection
+
+import (
+	"errors"
+	"os"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/uuid"
+)
+
+const ConfigMapNameEnv = "CONFIG_LEADERELECTION_NAME"
+
+var errEmptyLeaderElectionConfig = errors.New("empty leader election configuration")
+
+// NewConfigFromMap returns a Config for the given map, or an error.
+func NewConfigFromMap(data map[string]string) (*Config, error) {
+	config := defaultConfig()
+
+	if resourceLock, ok := data["resourceLock"]; ok {
+		config.ResourceLock = resourceLock
+	}
+
+	if leaseDurationStr, ok := data["leaseDuration"]; ok {
+		if leaseDuration, err := time.ParseDuration(leaseDurationStr); err == nil {
+			config.LeaseDuration = leaseDuration
+		}
+	}
+
+	if renewDeadlineStr, ok := data["renewDeadline"]; ok {
+		if renewDeadline, err := time.ParseDuration(renewDeadlineStr); err == nil {
+			config.RenewDeadline = renewDeadline
+		}
+	}
+
+	if retryPeriodStr, ok := data["retryPeriod"]; ok {
+		if retryPeriod, err := time.ParseDuration(retryPeriodStr); err == nil {
+			config.RetryPeriod = retryPeriod
+		}
+	}
+
+	if enabledComponents, ok := data["enabledComponents"]; ok {
+		tokens := strings.Split(enabledComponents, ",")
+		config.EnabledComponents = sets.NewString(tokens...)
+	}
+
+	return &config, nil
+}
+
+// NewConfigFromConfigMap returns a new Config from the given ConfigMap.
+func NewConfigFromConfigMap(configMap *corev1.ConfigMap) (*Config, error) {
+	if configMap == nil {
+		config := defaultConfig()
+		return &config, nil
+	}
+
+	return NewConfigFromMap(configMap.Data)
+}
+
+// Config represents the leader election config for a set of components
+// contained within a single namespace. Typically these will correspond to a
+// single source repository, viz: serving or eventing.
+type Config struct {
+	ResourceLock      string
+	LeaseDuration     time.Duration
+	RenewDeadline     time.Duration
+	RetryPeriod       time.Duration
+	EnabledComponents sets.String
+}
+
+func (c *Config) GetComponentConfig(name string) ComponentConfig {
+	if c.EnabledComponents.Has(name) {
+		return ComponentConfig{
+			LeaderElect:   true,
+			ResourceLock:  c.ResourceLock,
+			LeaseDuration: c.LeaseDuration,
+			RenewDeadline: c.RenewDeadline,
+			RetryPeriod:   c.RetryPeriod,
+		}
+	}
+
+	return defaultComponentConfig()
+}
+
+func defaultConfig() Config {
+	return Config{
```

Comment:
```suggestion
	return &Config{
```
This way you don't have to take address of the returned object every time